rfc/rfc-60/rfc-60.md

```diff
@@ -0,0 +1,225 @@
+This would help distribute the requests evenly across different prefixes, resulting in Amazon S3 creating partitions for
+the prefixes each with its own request limit. This significantly reduces the possibility of hitting the request limit
+for a specific prefix/partition.
+
+In addition, we want to expose an interface that provides users the flexibility to implement their own strategy for
+distributing files if using the traditional Hive storage strategy or federated storage strategy (proposed in this RFC) does
+not meet their use-case.
+
+## Design
+
+### Interface
+
+```java
+/**
+ * Interface for providing storage file locations.
+ */
+public interface HoodieStorageStrategy extends Serializable {
+  /**
+   * Return a fully-qualified storage file location for the given filename.
+   *
+   * @param fileId data file ID
+   * @return a fully-qualified location URI for a data file
+   */
+  String storageLocation(String fileId);
+
+  /**
+   * Return a fully-qualified storage file location for the given partition and filename.
+   *
+   * @param partitionPath partition path for the file
+   * @param fileId data file ID
+   * @return a fully-qualified location URI for a data file
+   */
+  String storageLocation(String partitionPath, String fileId);
+}
+```
+
+### Generating file paths for object store optimized layout
+
+We want to distribute files evenly across multiple random prefixes, instead of following the traditional Hive storage
+layout of keeping them under a common table path/prefix. In addition to the `Table Path`, for this new layout user will
+configure another `Table Storage Path` under which the actual data files will be distributed. The original `Table Path` will
+be used to maintain the table/partitions Hudi metadata.
+
+For the purpose of this documentation lets assume:
+```
+Table Path => s3://<table_bucket>/<hudi_table_name>/
+
+Table Storage Path => s3://<table_storage_bucket>/
+```
+Note: `Table Storage Path` can be a path in the same Amazon S3 bucket or a different bucket. For best results,
+`Table Storage Path` should be a top-level bucket instead of a prefix under the bucket to avoid multiple 
+tables sharing the prefix.
+
+We will use a Hashing function on the `Partition Path/File ID` to map them to a prefix generated under `Table Storage Path`:
+```
+// Hashing on the file ID makes sure that base file and its log files fall under the same folder
+
+s3://<table_storage_bucket>/<hash_prefix>/..
+```
+
+In addition, under the hash prefix we will follow a folder structure by appending Hudi Table Name and Partition. This
+folder structuring would be useful if we ever have to do a file system listing to re-create the metadata file list for
+the table (discussed more in the next section). Here is how the final layout would look like for `partitioned` tables:
+```
+s3://<table_storage_bucket>/01f50736/<hudi_table_name>/country=usa/075f3295-def8-4a42-a927-07fd2dd2976c-0_7-11-49_20220301005056692.parquet
```

Comment:
I didn't understand why using a random prefix before the table name which complicates all the operations (listing, cleaning, ...) while we can group the files under a common prefix, by removing a part of the file name and adding it to the prefix to make sure that there are no two files with the same prefix :
```
s3://<table_storage_bucket>/<hudi_table_name>/<partition1>/.../<partition N>/<commit timestamp>/<write token>/<file id>.parquet
s3://<table_storage_bucket>/<hudi_table_name>/<partition1>/.../<partition N>/<commit timestamp>/<a different write token>/<file id>.log.1_0-22-26
```
knowing that in S3, prefixes are considered to be the whole path of an object's location and not only the first X characters in the prefix, and the reading/writing quotas are per full prefix.
With this pattern, we can easily delete the data for a table, delete a partition, and listing and deleting the files of a commit or the versions of a file are much easier than the proposed pattern. In addition, the implementation is much easier because we just shift a part of the file name to the prefix.

Reply:
Based on the glossary here: https://docs.aws.amazon.com/general/latest/gr/glos-chap.html#keyprefix 
The prefix can be any length, so in the example given here hot partition or skewed partition would still be an issue